Mathlib/Algebra/MvPolynomial/Module/Basic.lean

```diff
@@ -0,0 +1,425 @@
+  rw [← monomial_zero', ← monomial_single_add, add_zero]
+
+theorem X_smul_C_eq_monomial {i : σ} {m : M} :
+    (X i : MvPolynomial σ R) • C R m = monomial R (Finsupp.single i 1) m := by
+  rw [← X_pow_smul_C_eq_monomial, pow_one]
+
+@[simp]
+theorem monomial_smul_monomial (a : σ →₀ ℕ) (r : R) (b : σ →₀ ℕ) (m : M) :
+    MvPolynomial.monomial a r • monomial R b m = monomial R (a + b) (r • m) := by
+  conv_lhs => rw [← mul_one r]
+  simp only [← smul_eq_mul, map_smul, smul_assoc]
+  refine congrArg (r • ·) ?_
+  induction' a using Finsupp.induction with _ _ _ _ _ ih
+  · rw [MvPolynomial.monomial_zero', MvPolynomial.C_1, one_smul, zero_add]
+  · rw [MvPolynomial.monomial_single_add, mul_smul, ih, add_assoc, monomial_single_add]
+
+@[simp]
+lemma monomial_smul_C (a : σ →₀ ℕ) (r : R) (m : M) :
+    MvPolynomial.monomial a r • C R m = r • monomial R a m := by
+  simp_rw [C, monomial_smul_monomial, add_zero, map_smul]
+
+lemma prod_X_pow_smul_eq_monomial (a : σ →₀ ℕ) (m : M) :
+    (∏ i in a.support, X i ^ a i : MvPolynomial σ R) • C R m = monomial R a m := by
+  rw [prod_X_pow_eq_monomial, monomial_smul_C, one_smul]
+
+theorem monomial_eq {a : σ →₀ ℕ} {m : M} :
+    monomial R a m = (a.prod fun n e => X n ^ e : MvPolynomial σ R) • C R m :=
+  (prod_X_pow_smul_eq_monomial a m).symm
+
+@[simp]
+theorem support_sum_monomial_coeff (p : MvPolynomialModule σ R M) :
+    (∑ v in p.support, monomial R v (coeff v p)) = p :=
+  Finsupp.sum_single p
+
+theorem as_sum (p : MvPolynomialModule σ R M) :
+    p = ∑ v in p.support, monomial R v (coeff v p) :=
+  (support_sum_monomial_coeff p).symm
+
+variable (σ R)
+
+open Submodule in
+lemma span_C_over_poly_eq_span_monomials_over_base (s : Set M) :
+    restrictScalars R (span (MvPolynomial σ R) (C R '' s)) =
+      ⨆ a : σ →₀ ℕ, map (monomial R a) (span R s) := by
+  simp_rw [map_span]
+  let K := ⨆ a : σ →₀ ℕ, span R (monomial R a '' s)
+  let N := span (MvPolynomial σ R) ((C R : M → MvPolynomialModule σ R M) '' s)
+  have H (f : MvPolynomial σ R) : K ≤ comap (Module.toModuleEnd R _ f) K := by
+    simp_rw [K, ← map_le_iff_le_comap, Submodule.map_iSup, iSup_le_iff, ← map_span]
+    intro a
+    induction' f using MvPolynomial.induction_on' with b r _ _ H₁ H₂
+    · refine le_iSup_of_le (b + a) ?_
+      refine le_trans (le_of_eq ?_) (smul_le_self_of_tower r _)
+      refine Eq.trans (map_comp _ _ _).symm <| Eq.trans ?_ (map_comp _ _ _)
+      congr 1; ext : 1
+      exact Eq.trans (monomial_smul_monomial _ _ _ _) (map_smul _ r _)
+    · rw [map_add]
+      exact le_trans (Submodule.map_add_le _ _ _) (sup_le H₁ H₂)
+  refine le_antisymm (?_ : N ≤ ⟨K.toAddSubmonoid, H⟩) ?_
+  · exact span_le.mpr (fun _ h => mem_iSup_of_mem 0 (subset_span h))
+  · refine iSup_le fun a => span_le.mpr (image_subset_iff.mpr ?_)
+    intro m h
+    refine (congrArg (· ∈ span _ _) monomial_eq).mpr ?_
+    exact smul_mem _ _ (subset_span (mem_image_of_mem _ h))
+
+variable (M)
+
+lemma top_eq_span_base_monomials :
```

Comment:
Similarly this is maybe `iSup_range_monomial_eq_top`, and the direction of the lemma should be reversed.